src/transformers/modeling_bart.py

```diff
@@ -0,0 +1,970 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""PyTorch BART model, ported from the fairseq repo."""
+
+import logging
+import random
+from collections import namedtuple
+from typing import Dict, List, Optional, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor, nn
+
+from .configuration_bart import BartConfig
+from .file_utils import add_start_docstrings
+from .modeling_utils import PreTrainedModel
+
+
+logger = logging.getLogger(__name__)
+
+
+BART_PRETRAINED_MODEL_ARCHIVE_MAP = {
+    "bart-large": "https://s3.amazonaws.com/models.huggingface.co/bert/bart-large-pytorch_model.bin",
+    "bart-large-mnli": "https://s3.amazonaws.com/models.huggingface.co/bert/bart-large-mnli-pytorch_model.bin",
+}
+
+
+BART_START_DOCSTRING = r"""  TODO(SS): FIXME"""
+
+
+class PretrainedBartModel(PreTrainedModel):
+    config_class = BartConfig
+    base_model_prefix = "model"
+    pretrained_model_archive_map = BART_PRETRAINED_MODEL_ARCHIVE_MAP
+
+    def _init_weights(self, module):
+        std = self.init_std  # used by init_params
+
+        # called init_bert_params in fairseq
+        if isinstance(module, nn.Linear):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.bias is not None:
+                module.bias.data.zero_()
+        if isinstance(module, nn.Embedding):
+            module.weight.data.normal_(mean=0.0, std=std)
+            if module.padding_idx is not None:
+                module.weight.data[module.padding_idx].zero_()
+
+
+# Public API
+
+
+@add_start_docstrings(
+    "The bare BART Model model outputting raw hidden-states without any specific head on top.",
+    BART_START_DOCSTRING,
+    "FIXME(SS)",
+)
+class BartModel(PretrainedBartModel,):
+    """FIXME(SS)"""
+
+    def __init__(self, config: BartConfig):  # should take config
+        super().__init__(config)
+        self.config = config
+        self._is_generation_fast = False
+        self.output_attentions = config.output_attentions
+        self.output_hidden_states = config.output_hidden_states
+
+        padding_idx, vocab_size = config.pad_token_id, config.vocab_size
+        self.shared = nn.Embedding(vocab_size, config.d_model, padding_idx)
+
+        self.encoder = BartEncoder(config, self.shared)
+        self.decoder = BartDecoder(config, self.shared)
+
+        self.init_std = config.init_std
+        self.init_weights()
+        self._is_generation_fast = False  # TODO(SS): this might need deletion
+
+    def get_input_embeddings(self):
+        return self.shared
+
+    def set_input_embeddings(self, value):
+        self.shared = value
+
+    def forward(self, input_ids: torch.LongTensor = None, return_for_head=False, **kwargs):
```

Comment:
> I think we should have all arguments being provided with either an encoder_ prefix if the related argument will be fed to the encoder or a decoder_ prefix if the related arguments will be fed to the decoder.



My preferred signature for Seq2Seq APIs:
```
    def forward(self, input_ids=None,  # encoder_input_ids
                attention_mask=None, # encoder_attention_mask
                head_mask=None,  # again for encoder
                encoder_hidden_states=None,  # or name this encoder_output
                decoder_input_ids=None,  # sent to decoder
                decoder_attention_mask=None,  # sent to decoder
                ):

```
This allows the caller to only engage with the seq2seq nature of the model if they want to, e.g:
- `BartForSequenceClassification()(tokens)`
- `BartForSequenceClassification()(input_ids=tokens)`

I agree it is a little bit less explicit, so I'm happy to do it either way.

Reply:
Seems like a lot of design choices were made around Summarization, where we definitely need `decoder_input_ids`.

Reply:
I'm not a fan of allowing `the caller to ignore the seq2seq nature of the model`. In the library, we try to be very explicit on the inner working of the models and I feel like designing the encoder-decoder API is a place where we should be careful to make the right decision so that they are not confused with single-stack models.

I know the Bart model duplicates the input for sequence classification but I think that this is mostly a temporary trick for sequence classification so I don't want this to influence too strongly how we shape the encoder-decoder API in general (for instance, T5 has a different approach). This is related to my comment [here](https://github.com/huggingface/transformers/pull/2745#discussion_r378227300) (the other very important part to discuss).

Here is what I would propose:
- let's have a very clean and normal forward loop in the main encoder-decoder `BartModel`, i.e. with explicit `encoder_input_ids` and `decoder_input_ids` and no duplication of `input_ids` inside the forward loop.
- implement the duplication trick of the inputs for both encoder and decoder in the forward loop of `BartForSequenceClassification`.

What do you think?

Reply:
Its not just for classification, `BartForMaskedLM` also uses it. 

So now I have the mask making up front on 
on `BartModel`.
if attention_mask is None: make it  # for encoder to ignore padding keys
if decoder_input_ids is None: make it
if decoder_attention_mask is None: make it

This obviously goes against the philosophy of what you said, so I wanted to check if knowing that the trick is also used for pretraining changes your view. (I think UniLM is similar)

Reply:
I see, let me investigate a bit more. Can you link to the relevant original code?
Also, what do you mean by "So now I have the mask making up front on
on BartModel."?